apps/studio/src/features/permissions/PermissionsContext.tsx

```diff
@@ -0,0 +1,58 @@
+import type { PropsWithChildren } from "react"
+import { createContext, useContext } from "react"
+import { AbilityBuilder, createMongoAbility } from "@casl/ability"
+import { RoleType } from "@prisma/client"
+
+import type { ResourceAbility } from "~/server/modules/permissions/permissions.type"
+import { buildPermissionsFor } from "~/server/modules/permissions/permissions.util"
+import { trpc } from "~/utils/trpc"
+
+interface PermissionsContextReturn {
+  ability: ResourceAbility
+}
+
+interface PermissionsProviderProps {
+  siteId: number
+  resourceId?: string
+}
+
+const getPermissions = (roles: { role: RoleType }[]) => {
+  const builder = new AbilityBuilder<ResourceAbility>(createMongoAbility)
+  roles.forEach(({ role }) => {
+    buildPermissionsFor(role, builder)
+  })
+  return builder.build({ detectSubjectType: () => "Resource" })
+}
+
+export const PermissionsContext =
+  createContext<PermissionsContextReturn | null>(null)
+
+export const PermissionsProvider = ({
+  children,
+  siteId,
+  resourceId,
+}: PropsWithChildren<PermissionsProviderProps>) => {
+  const [roles] = trpc.resource.getRolesFor.useSuspenseQuery({
+    siteId,
+    resourceId: resourceId ?? null,
+  })
+
+  // TODO: need to get the roles from backend
```

Comment:
isn't the roles coming from BE here?

Reply:
you're right, i had this comment previously to remind myself oops. removed!